java-checks/src/main/java/org/sonar/java/checks/SwitchCaseWithoutBreakCheck.java

```diff
@@ -21,66 +21,65 @@
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Lists;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.regex.Pattern;
+import java.util.stream.IntStream;
 import org.sonar.check.Rule;
 import org.sonar.java.ast.visitors.SubscriptionVisitor;
-import org.sonar.plugins.java.api.JavaFileScanner;
-import org.sonar.plugins.java.api.JavaFileScannerContext;
-import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
-import org.sonar.plugins.java.api.tree.BlockTree;
+import org.sonar.java.cfg.CFG;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
 import org.sonar.plugins.java.api.tree.CaseGroupTree;
-import org.sonar.plugins.java.api.tree.CaseLabelTree;
-import org.sonar.plugins.java.api.tree.IfStatementTree;
-import org.sonar.plugins.java.api.tree.StatementTree;
 import org.sonar.plugins.java.api.tree.SwitchStatementTree;
 import org.sonar.plugins.java.api.tree.SyntaxTrivia;
 import org.sonar.plugins.java.api.tree.Tree;
-import org.sonar.plugins.java.api.tree.TryStatementTree;
-
-import java.util.Collections;
-import java.util.List;
-import java.util.regex.Pattern;
 
 @Rule(key = "S128")
-public class SwitchCaseWithoutBreakCheck extends BaseTreeVisitor implements JavaFileScanner {
-
-  private JavaFileScannerContext context;
-
+public class SwitchCaseWithoutBreakCheck extends IssuableSubscriptionVisitor {
   @Override
-  public void scanFile(JavaFileScannerContext context) {
-    this.context = context;
-    scan(context.getTree());
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.SWITCH_STATEMENT);
   }
 
   @Override
-  public void visitSwitchStatement(SwitchStatementTree switchStatement) {
-    switchStatement.cases().stream()
-      // Exclude the last case as stated in RSPEC. This also excludes switches with no or a single case group.
-      .limit(Math.max(0, switchStatement.cases().size() - 1))
-      .forEach(caseGroup -> {
-        // Assign issues to the last label in the group
-        CaseLabelTree caseLabel = caseGroup.labels().get(caseGroup.labels().size() - 1);
+  public void visitNode(Tree tree) {
+    SwitchStatementTree switchStatementTree = (SwitchStatementTree) tree;
+    List<CaseGroupTree> caseGroupTrees = switchStatementTree.cases();
+    CFG cfg = CFG.buildCFG(Collections.singletonList(tree), true);
+    CFG.Block entry = cfg.entry();
+    List<CFG.Block> cfgCases = new ArrayList<>(entry.successors());
+
+    if (!hasDefaultClause(switchStatementTree)) {
+      cfgCases.remove(cfgCases.size() - 1);
```

Comment:
`CFG.Block.successors` returns a `Set<Block>` and I don't think that's any guarantee on the order of the `case` blocks.